models/cleanup.go

```diff
@@ -0,0 +1,118 @@
+package models
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/netlify/gotrue/storage"
+	"github.com/sirupsen/logrus"
+	metricglobal "go.opentelemetry.io/otel/metric/global"
+	metricinstrument "go.opentelemetry.io/otel/metric/instrument"
+	otelasyncint64instrument "go.opentelemetry.io/otel/metric/instrument/asyncint64"
+)
+
+var cleanupStatements []string
+var cleanupAffectedRows otelasyncint64instrument.Counter
+
+func init() {
+	tableRefreshTokens := RefreshToken{}.TableName()
+	tableSessions := Session{}.TableName()
+
+	// These statements intentionally use a FOR UPDATE SKIP LOCKED clause
+	// in the select portion which ensures that the transaction performing
+	// the cleanup does not wait on other transactions on the database that
+	// are touching those rows. This ensures that cleanups always run as
+	// fast as possible while consuming minimal resources to do it.
+	cleanupStatements = append(cleanupStatements,
+		// deletes 100 refresh tokens without waiting on locked rows
+		fmt.Sprintf("delete %s where id in (select id from %s where revoked is true and updated_at < (now() - interval '10 minutes' limit 100 for update skip locked);", tableRefreshTokens, tableRefreshTokens),
```

Comment:
Also, i'm not sure if there's a point using `skip locked` here because the `delete` statement will need to hold a write lock anyway so if there's another transaction that's holding on to the write lock, we need to wait. 

In terms of performance, can we use a CTE instead of a correlated subquery? IIRC a CTE will only be invoked once while a correlated subquery will be invoked on every row.

Reply:
Yeah am a little confused here as well - Not too familiar with the locking system in Postgres - would the perf gains come from reduced waiting time on row level locks?

My current understanding(assuming default read committed txn mode) is:

1. Table level locks for a Delete statement can [run with other delete statements](https://www.citusdata.com/blog/2018/02/15/when-postgresql-blocks/)
<img width="653" alt="CleanShot 2023-01-03 at 12 15 27@2x" src="https://user-images.githubusercontent.com/8011761/210299362-8633bc34-62aa-459c-a314-530e0a793378.png">

2. For row level locks [`for update` row level lock mode will be taken on both `delete` and `update`](https://www.postgresql.org/docs/current/explicit-locking.html) which means that 
> other transactions that attempt UPDATE, DELETE, SELECT FOR UPDATE, SELECT FOR NO KEY UPDATE, SELECT FOR SHARE or SELECT FOR KEY SHARE of these rows will be blocked

Would the savings come from being able to execute some of the deletes first while the other rows are being updated? Contrived example below:

txn1: 
`update set name='harry' from contacts where idx=1`

txn2:
`delete * from contacts`

Table: `contacts`

idx | name | phone | metadata
1    | John  | 12345 | stuff
2    | Bob   | 12346 | rad
3    | Billy   | 12347 | asd

so if these 2 `txns` are run concurrently then Postgres can 
1. delete rows 2 and 3, 
2. update row 1
3. delete row 1

Reply:
> Also, i'm not sure if there's a point using `skip locked` here because the `delete` statement will need to hold a write lock anyway so if there's another transaction that's holding on to the write lock, we need to wait.

The point with `skip locked` is to acquire a write lock on rows that are not currently used by some other process. The idea is that per-request cleanups are least priority and should only do the minimal amount of work without disturbing other more important processes. 

This also should answer some of the questions by @J0.